docs/guides/validation-user-guide.md

```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
 bodyclass: docs
 layout: docs
 ---
-<p class="coming-soon">Coming soon...</p>
+
+Spine uses Protobuf for defining domain models. It also helps us to standardize creating
+[Value Objects](https://martinfowler.com/bliki/ValueObject.html). Both those usages require
+something more than just a data structure. One of the commodities required for describing domain
+specifics is data validation. Spine builds on top of Protobuf with a Validation library.
+
+The purpose of the Validation library is:
+ - giving framework users an easy way to maintain simple domain invariants;
+ - giving framework users an easy way to construct simple Anticorruption layers.
+
+This guide will walk you though the API of Spine Validation library. All of the validation features
+described here are currently supported in the Java environment. Many are supported in Dart as well.
+For more info, see the description of individual constraints.
+
+## Validation options
+### Required fields
+
+In Protobuf 2 the concept of required fields used to be built into the language. This proved to be
+a [design mistake](https://stackoverflow.com/a/31814967/3183076) and in Protobuf 3 all the fields
+are optional. However, we've revived the concept in Spine Validation library. Unlike the Protobuf 2
+`required` fields, our `(required)` fields do not effect message serialization and thus can be
+ignored. Rather than guarding a technical invariant, `(required)` guards a domain model invariant.
+The choice whether or not to use validation lies solely on the developer.
+
+#### How required fields work
+
+Fields in Protobuf may have either a primitive type or a user-defined type. A user-defined type is
```

Comment:
Previously I commented on the flow "simple definition" -> "simple positive example" -> "more details" -> "another example" -> "negative scenarios and don'ts with more technical stuff".

It is still valid as of now.